be/src/exec/pipeline/crossjoin/nljoin_probe_operator.cpp

```diff
@@ -0,0 +1,384 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "exec/pipeline/crossjoin/nljoin_probe_operator.h"
+
+#include "column/chunk.h"
+#include "column/column_helper.h"
+#include "exec/pipeline/crossjoin/cross_join_right_sink_operator.h"
+#include "gen_cpp/PlanNodes_types.h"
+#include "runtime/descriptors.h"
+#include "simd/simd.h"
+#include "storage/chunk_helper.h"
+
+namespace starrocks::pipeline {
+
+NLJoinProbeOperator::NLJoinProbeOperator(OperatorFactory* factory, int32_t id, int32_t plan_node_id,
+                                         int32_t driver_sequence, TJoinOp::type join_op,
+                                         const std::string& sql_join_conjuncts,
+                                         const std::vector<ExprContext*>& join_conjuncts,
+                                         const std::vector<ExprContext*>& conjunct_ctxs,
+                                         const std::vector<SlotDescriptor*>& col_types, size_t probe_column_count,
+                                         size_t build_column_count,
+                                         const std::shared_ptr<CrossJoinContext>& cross_join_context)
+        : OperatorWithDependency(factory, id, "nestloop_join_probe", plan_node_id, driver_sequence),
+          _join_op(join_op),
+          _col_types(col_types),
+          _probe_column_count(probe_column_count),
+          _build_column_count(build_column_count),
+          _sql_join_conjuncts(sql_join_conjuncts),
+          _join_conjuncts(join_conjuncts),
+          _conjunct_ctxs(conjunct_ctxs),
+          _cross_join_context(cross_join_context) {
+    _cross_join_context->ref();
+}
+
+Status NLJoinProbeOperator::prepare(RuntimeState* state) {
+    _runtime_state = state;
+    _output_accumulator.set_desired_size(state->chunk_size());
+
+    _unique_metrics->add_info_string("join_conjuncts", _sql_join_conjuncts);
+    return Operator::prepare(state);
+}
+
+void NLJoinProbeOperator::close(RuntimeState* state) {
+    _cross_join_context->unref(state);
+    Operator::close(state);
+}
+
+bool NLJoinProbeOperator::is_ready() const {
+    return _cross_join_context->is_right_finished();
+}
+
+bool NLJoinProbeOperator::_is_curr_probe_chunk_finished() const {
+    return _probe_chunk == nullptr || _probe_row_current >= _probe_chunk->num_rows();
+}
+
+void NLJoinProbeOperator::_advance_join_stage(JoinStage stage) const {
+    DCHECK_LE(_join_stage, stage) << "current=" << _join_stage << ", advance to " << stage;
+    if (_join_stage != stage) {
+        _join_stage = stage;
+        VLOG(3) << fmt::format("operator {} enter join_stage {}", _driver_sequence, stage);
+    }
+}
+
+void NLJoinProbeOperator::_check_post_probe() const {
+    bool output_finished = _is_curr_probe_chunk_finished() && _output_accumulator.empty();
+    if (_input_finished && output_finished) {
```

Comment:
when build side is empty, the probe side(inner/right/cross join) should finished earlier in short-circuit way in some situations,

Reply:
done